pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -49,11 +53,54 @@ type ReaderPool struct {
 	// Keep track of all readers managed by the pool.
 	lazyReadersMx sync.Mutex
 	lazyReaders   map[*LazyBinaryReader]struct{}
+	tracker       HeadersLazyLoadedTracker
+}
+
+// HeadersLazyLoadedTrackerState tracks blocks that are lazy loaded and its last access time.
+type HeadersLazyLoadedTrackerState struct {
+	sync.Mutex
+	state storepb.IndexHeaderTrackerState
+}
+
+// HeadersLazyLoadedTracker persists the list of lazy loaded blocks.
+type HeadersLazyLoadedTracker struct {
+	lazyReaders map[*LazyBinaryReader]struct{}
+	userID      string
+	persistPath string
+}
+
+// Write writes the list of lazy loaded block to the writer.
+func (h *HeadersLazyLoadedTracker) Write() error {
+	trackerState := HeadersLazyLoadedTrackerState{
+		state: storepb.IndexHeaderTrackerState{
+			LazyLoadedBlocks: make(map[string]uint32),
+			UserId:           h.userID,
+		},
+	}
+
+	trackerState.Lock()
+	defer trackerState.Unlock()
```

Comment:
I'm not sure what this mutex is protecting against. I think it's safe to remote it. At that point you should also be able to remove `HeadersLazyLoadedTrackerState` too and only use `storepb.IndexHeaderTrackerState`